internal/client/client.go

```diff
@@ -88,11 +88,33 @@ func (c *Config) Validate() error {
 	return nil
 }
 
-type ShutdownHandler func(ctx context.Context)
+type ShutdownHandlerType int
+
+type ShutdownHandler map[ShutdownHandlerType]func(ctx context.Context)
+
+const (
+	ProxyShutdownHandler ShutdownHandlerType = iota
+	NearCacheShutdownHandler
+)
+
+func (shutdownHandler *ShutdownHandler) AddShutdownHandler(handlerID ShutdownHandlerType, handler func(ctx context.Context)) {
+	// this is supposed to be called during client initialization, so there's no risk of races.
+	if _, ok := (*shutdownHandler)[handlerID]; ok {
+		panic(fmt.Errorf("handlerID: %d was already added to shutdown handler before", handlerID))
+	}
+	(*shutdownHandler)[handlerID] = handler
+}
+
+func (shutdownHandler *ShutdownHandler) executeShutdownHandlers(ctx context.Context) {
+	for _, f := range *shutdownHandler {
+		f(ctx)
+	}
+}
```

Comment:
Are these necessary? How about adding the following method:
```
func (c *Client) destroyProxies(ctx context.Context) {
  c.proxyManager.destroyProxies(ctx)
}
```
And then:
```
c.ic.AddShutdownHandler(c.destroyProxies)
```

Reply:
I think that these are the necessary from the perspective of **readability** and **maintenance**. But I can it can be introducing additional destroyProxies to client method set. [df88887](https://github.com/hazelcast/hazelcast-go-client/pull/772/commits/df8888728cd4a5f4a040179d4df2b0cb77c1c8b6).

Reply:
I think the complexity here is making it less readable and maintainable. Do we need to have different types of shutdown handlers?

Reply:
Simplied the design, [694aee3](https://github.com/hazelcast/hazelcast-go-client/pull/772/commits/694aee3b72c6dd7465c5d3bde999c2fdc292f28e).